### PR TITLE
Toric Varieties/Schemes: Ep/toric blowups center #3810

### LINKED ISSUE #3810: Center of a blowup
There has been some discussion about what the method `center` of a blowup should return. 
In personal discussion with @HechtiDerLachs  and @paemurru  we came to the following:

- mathematically speaking the center  is a reduced subscheme, there is no canonical non-reduced structure of it.
Thus `center` must return a the reduced subscheme and the inclusion morphism. Since computing the reduced subscheme is impractical, for now we raise a `NotImplementedError`  until in the possibly far future we have a lazy type for reduced subschemes. 

There should be a method 
`center_ideal_sheaf` which returns the lazy radical ideal sheaf defining the center. This data structure exists already. 
Existing code that uses `center` can be redirected to `center_ideal_sheaf`. The method `center_ideal_sheaf` will be exported. 

Let me ping the other stakeholders @afkafkafk13 @HereAround 
Do you have suggestions / opinions on this?

If we all agree, I will ask @paemurru  to implement this.

### PR DESCRIPTION
**Pull request by paemurru**
Resolves https://github.com/oscar-system/Oscar.jl/issues/3810.

Small mathematical, documentation and name changes in toric blowups.

@lkastner @HereAround @HechtiDerLachs @afkafkafk13 Please have a look.

(1) `center` -> `center_unnormalized`

Given a toric blowup `f`, the `center_unnormalized(f)` (previously called `center(f)` in ToricBlowups) computes an ideal sheaf `I` such that if we blow up along `I` and normalize, we get the domain of `f`.

It is easy to compute and is related to the reduced center from the works of Dan Abramovich, Michael Temkin, Jarosław Włodarczyk (which despite its name, corresponds to a certain, usually nonreduced, subscheme).

Finding the center-in-the-sense-of-the-center-of-a-birational-morphism is more computational as it involves computing the radical. Also, computing the center-in-the-sense-of-the-center-of-a-blowup can be computational, as it involves taking a suitable multiple of the prime divisor, equivalently, a Veronese subalgebra of possibly high degree. These have not been currently added.

(2) `exceptional_divisor` -> `exceptional_prime_divisor`

Given a toric blowup `f` adding exactly one ray (which may coincide with a previous one), the `exceptional_prime_divisor(f)` (previously called `exceptional_divisor(f)` in ToricBlowups) computes the prime divisor `D` that corresponds to the selected ray. It is not necessarily the exceptional locus of a birational morphism and it is not necessarily the effective Cartier divisor of a blowup.

The function `exceptional_divisor` previously failed when called on projective 3-space adding a ray with coordinates greater than 1. Fixes this now.

(3) Add `center_data`

@afkafkafk13 wished to have a getter for the initial data used to construct the blowup. This is provided by `center_data`.

(4) Write documentation for the above, also clarifications in markdown files

(5) Extra type parameters in ToricBlowdownMorphism type

To do multiple dispatch in a type stable way (https://github.com/oscar-system/Oscar.jl/issues/3810#issuecomment-2144580976), added `CenterDataType` and `CenterUnnormalizedType` as type parameters.

(6) Comment on type instability

At the moment, the interfaces in ToricBlowups and BlowupMorphism are rather different. The exceptional divisor is a prime divisor in ToricBlowups but is an effective Cartier divisor in BlowupMorphism. The center is an unnormalized center (as explained above) in ToricBlowups but is the center-of-a-blowup (as explained above) in BlowupMorphism. Also, depending on the type of `center_unnormalized`, it is currently sometimes possible to turn it into an ideal in the Cox ring, sometimes not.

The constructor `blow_up` is currently type unstable. Due to the above points, it is important for the user to readily see the types and have clarity on which functions can change the type and in what ways. I have tried to document this and make it explicit.

### CODE REVIEW

**Comment by afkafkafk13** [comment]
> The constructor `blow_up` is currently type unstable. Due to the above points, it is important for the user to readily see the types and have clarity on which functions can change the type and in what ways. I have tried to document this and make it explicit.

If `blow_up` is called with arguments clearly belonging to the toric world, it should stay there; if it is called with arguments in the scheme world, it should stay in this world. If the input it ambiguous (which should rarely happen), the user needs a variant to force it into one world or the other.

File: experimental/Schemes/src/ToricBlowups/attributes.jl [hunk 1]
```diff
@@ -75,43 +103,45 @@ function center(bl::ToricBlowdownMorphism)
...
     X = domain(bl)
     S = cox_ring(X)
     x = gens(S)
     j = index_of_new_ray(bl)
     help_list = [i == j ? 1 : 0 for i in 1:ngens(S)]
     td = toric_divisor(X, help_list)
-    @assert is_cartier(td) "exceptional divisor must be Cartier"
-    @assert is_prime(td) "exceptional divisor must be prime"
-    bl.exceptional_divisor = td
+    @assert is_prime(td) "exceptional prime divisor must be prime"
```
Comment (afkafkafk13) [inline experimental/Schemes/src/ToricBlowups/attributes.jl:147]: This does not seem very helpful -- if it is an exceptional prime divisor by type, then it is forcibly prime and we do not need to assert it. If the assert is necessary, then the prime divisor type has a problem....

File: experimental/Schemes/src/ToricBlowups/constructors.jl [hunk 2]
```diff
@@ -1,7 +1,10 @@
 @doc raw"""
     blow_up(m::NormalToricVariety, I::ToricIdealSheafFromCoxRingIdeal; coordinate_name::String = "e")
 
-Blow up the toric variety along the center given by a toric ideal sheaf.
+Blow up the toric variety along a toric ideal sheaf.
+
+!!! warning
+    This function is type unstable. The type of the domain of the output `f` is always a subtype of `AbsCoveredScheme` (meaning that `domain(f) isa AbsCoveredScheme` is always true). Sometimes, the type of the domain will be a toric variety (meaning that `domain(f) isa NormalToricVariety` is true) if the algorithm can successfully detect this. In the future, the detection algorithm may be improved so that this is successful more often.
```
Comment (afkafkafk13) [inline experimental/Schemes/src/ToricBlowups/constructors.jl:7]: I already commented on the need for type stability in a separate comment.

**Review by afkafkafk13 — COMMENTED** [review]
Looks solid. I left you a few comments.

File: experimental/FTheoryTools/src/AbstractFTheoryModels/methods.jl [hunk 3]
```diff
@@ -72,10 +72,7 @@ julia> x1, x2, x3, x4, x, y, z = gens(cox_ring(ambient_space(t)))
 7-element Vector{MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}}:
  x1
  x2
- x3
- x4
- x
- y
+ ⋮
```
Comment (lgoettgens) [inline experimental/FTheoryTools/src/AbstractFTheoryModels/methods.jl]: These changes are unfortunately a bottleneck of  `doctest_fix`, which makes use of your terminal size for abbreviating lists. The doctests will fail with this, so I am afraid you have to revert this change (and the one a bit further down in this file)

File: experimental/Schemes/src/ToricBlowups/constructors.jl [hunk 4]
```diff
@@ -61,45 +64,52 @@ end
...
-a custom variable name.
+the exceptional prime divisor. As optional argument one can supply a
+custom variable name.
+
+!!! warning
+    This function is type unstable. The type of the field `center_unnormalized` is always a subtype of `AbsIdealSheaf` (meaning that `center_unnormalized(f) isa Oscar.AbsIdealSheaf` is always true). Sometimes, the function computes and sets the field `center_unnormalized` for the output `f`, giving it the type `ToricIdealSheafFromCoxRingIdeal` (meaning that `center_unnormalized(f) isa Oscar.ToricIdealSheafFromCoxRingIdeal` is true and `center_unnormalized(f) isa IdealSheaf` is false). If it does not, then calling `center_unnormalized(f)` computes and sets the field `center_unnormalized` and it will have the type `IdealSheaf` (meaning that `center_unnormalized(f) isa Oscar.ToricIdealSheafFromCoxRingIdeal` is false and `center_unnormalized(f) isa IdealSheaf` is true).
 
 # Examples
+
+In the example below `f.center_unnormalized` has type `ToricIdealSheafFromCoxRingIdeal` and we can access the corresponding ideal in the Cox ring using `Oscar.ideal_in_cox_ring`.
```
Comment (lgoettgens) [inline experimental/Schemes/src/ToricBlowups/constructors.jl]:
```suggestion
In the example below `center_unnormalized(f)` has type `ToricIdealSheafFromCoxRingIdeal` and we can access the corresponding ideal in the Cox ring using `Oscar.ideal_in_cox_ring`.
```
wouldn't it make more sense to guide users to the getter function instead of the direct access?

File: experimental/Schemes/src/ToricBlowups/constructors.jl [hunk 5]
```diff
@@ -61,45 +64,52 @@ end
...
-  x3
+  x2^2
+  x3^3
 ```
-Notice that in the above example, the blowup center is not just an ideal sheaf.
-Rather, it is an ideal sheaf that knows its datum, in the form of an ideal,
-in the Cox ring. Sadly, we cannot always (at least not yet) compute such a datum.
-The following example demonstrates such a case.
+
+In the below example, `f.center_unnormalized` has type `IdealSheaf` and we cannot access the corresponding ideal in the Cox ring.
```
Comment (lgoettgens) [inline experimental/Schemes/src/ToricBlowups/constructors.jl]:
```suggestion
In the below example, `center_unnormalized(f)` has type `IdealSheaf` and we cannot access the corresponding ideal in the Cox ring.
```
see above

File: experimental/Schemes/src/ToricBlowups/constructors.jl [hunk 6]
```diff
@@ -61,45 +64,52 @@ end
...
 Blow up the toric variety by subdividing the fan of the variety with the
-provided new ray. Note that this ray must be a primitive element in the
-lattice Z^d, with d the dimension of the fan. This function returns the
-corresponding blowdown morphism.
+provided new ray. This function returns the corresponding morphism.
+
+Note that this ray must be a primitive element in the lattice Z^d that
+does not coincide with any existing ray, with d the dimension of the
+fan. In particular, it is currently impossible to blow up along a ray
+which corresponds to a non-Q-Cartier divisor.
```
Comment (lkastner) [inline experimental/Schemes/src/ToricBlowups/constructors.jl]:
```suggestion
```
#3926 fixes this.

**Review by lkastner — APPROVED** [review]

File: docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md [hunk 7]
```diff
@@ -2,50 +2,41 @@
...
 
-In general, resolutions need not be toric. Indeed, some of the functionality below requires
-fully-fledge schemes machinery, which -- as of this writing (October 2023) -- is still in
+In general, resolutions of toric varieties need not be toric. Indeed, some of the functionality below requires
+fully-fledge schemes machinery, which -- as of October 2023 -- is still in
 Oscar's experimental state. For this reason, the methods below should be considered experimental.
 
+We focus mainly on toric blowups given by a star subdivision of a polyhedral fan along a primitive vector, see 11.1 Star Subdivisions in [CLS11](@cite). Below, we refer to this new primitive
+vector as `new_ray`.
+The main constructor is the following
```
Comment (HereAround) [inline docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md]: This looks like a strange (most likely unintentional) line break. Truly a nitpick. Please ignore (or fix if you want).

File: docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md [hunk 8]
```diff
@@ -2,50 +2,41 @@
...
+For our most specialized blowup method, we focus on the n-th cone in the fan of the variety `v` in question.
+This cone need not be maximal. The ensuing star subdivision will subdivide this cone about its "diagonal"
 (the sum of all its rays). The result of this will always be a toric variety:
 ```@docs
 blow_up(v::NormalToricVariety, n::Int; coordinate_name::String = "e")
 ```
-More generally, we can provide a primitive element in the fan of the variety in question. The resulting
-star subdivision leads to a polyhedral fan, or put differently, the blow-up along this center is always toric:
+More generally, we can provide a primitive element in the fan of the variety in question and construct a toric morphism as in Section 11.1 Star Subdivisions in [CLS11](@cite). The resulting
+star subdivision leads to a polyhedral fan, or put differently, the blowup is always toric:
```
Comment (HereAround) [inline docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md]: Here is another such line break. I will stop here. Please take a look (if you care to fix this).

**Review by HereAround — APPROVED** [review]
Thank you for working on this @paemurru . We just discussed with @lkastner and we are all happy with this. So good to go!

File: experimental/Schemes/src/ToricBlowups/constructors.jl (hunk 6 above)
Comment (paemurru) [reply experimental/Schemes/src/ToricBlowups/constructors.jl]: Removed "does not coincide with any existing ray".

File: docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md (hunk 7 above)
Comment (paemurru) [reply docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md]: Fixed.

File: docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md (hunk 8 above)
Comment (paemurru) [reply docs/src/AlgebraicGeometry/ToricVarieties/BlowdownMorphisms.md]: Fixed.
